x-pack/plugins/maps/public/shared/layers/vector_layer.js

```diff
@@ -517,13 +519,34 @@ export class VectorLayer extends AbstractLayer {
     return [this._getMbPointLayerId(), this._getMbLineLayerId(), this._getMbPolygonLayerId()];
   }
 
+
+  _addJoinsToSourceTooltips(tooltipsFromSource) {
+    for (let i = 0; i < tooltipsFromSource.length; i++) {
+      const tooltipProperty = tooltipsFromSource[i];
+      const matchingJoins = [];
+      for (let j = 0; j < this._joins.length; j++) {
+        if (this._joins[j].getLeftFieldName() === tooltipProperty.getPropertyName()) {
+          matchingJoins.push(this._joins[j]);
+        }
+      }
+      if (matchingJoins.length) {
+        tooltipsFromSource[i] = new JoinTooltipProperty(tooltipProperty, matchingJoins);
+      }
+    }
+  }
+
+
   async getPropertiesForTooltip(properties) {
-    const tooltipsFromSource =  await this._source.filterAndFormatPropertiesToHtml(properties);
+
+    let allTooltips =  await this._source.filterAndFormatPropertiesToHtml(properties);
+    this._addJoinsToSourceTooltips(allTooltips);
+
+
     for (let i = 0; i < this._joins.length; i++) {
       const propsFromJoin = await this._joins[i].filterAndFormatPropertiesForTooltip(properties);
-      Object.assign(tooltipsFromSource, propsFromJoin);
+      allTooltips = [...allTooltips, ...propsFromJoin];
```

Comment:
Instead of recreating the array on each iteration, how about something more like this?

```
  async getPropertiesForTooltip(properties) {
    const sourceTooltipProperties =  await this._source.filterAndFormatPropertiesToHtml(properties);
    this._addJoinsToSourceTooltips(sourceTooltipProperties);

    // tooltip properties added by joins. For example, aggregation metrics
    const joinTooltipProperties = [];
    for (let i = 0; i < this._joins.length; i++) {
      const propsFromJoin = await this._joins[i].filterAndFormatPropertiesForTooltip(properties);
      joinTooltipProperties.push(...propsFromJoin);
    }
    return [...sourceTooltipProperties, ...joinTooltipProperties];
  }
```